### PR TITLE
Override host header in http.proxy with hostname from <connect>

### PR DESCRIPTION
**Pull request by msalavastru**
Overridden host header in HttpProxyServiceHandler and updated the k3po scripts for the ITests in http.proxy.

### CODE REVIEW

**Comment by msalavastru** [comment]
Retest this please

**Comment by msalavastru** [comment]
retest this please

**Comment by irina-mitrea-luxoft** [comment]
retest this please

**Comment by msalavastru** [comment]
This is the config file that I used to do the testing.

[gateway-config-minimal.xml](https://files.zenhub.io/564df1972d47521d4101db78)
Uploaded using [**ZenHub.io**](http://www.zenhub.io)

File: transport/http/src/main/java/org/kaazing/gateway/transport/http/bridge/filter/HttpResponseDecodingState.java [hunk 1]
```diff
@@ -257,7 +257,8 @@ protected DecodingState finishDecode(List<Object> childProducts,
 						// create the cookie name and value
 						DefaultHttpCookie cookie = new DefaultHttpCookie(
 								nvPair[0].trim());
-						cookie.setValue(nvPair[1].trim());
+						if (nvPair.length > 1)
+						    cookie.setValue(nvPair[1].trim());
```
Comment (cmebarrow) [inline transport/http/src/main/java/org/kaazing/gateway/transport/http/bridge/filter/HttpResponseDecodingState.java]: @msalavastru This fix seems fine. Where's the k3po test you mentioned that verifies this fix? Also, as mentioned in our call just now, please add a test method to unit test HttpResponseDecoderTest to confirm the fix (make sure it fails without the fix).

File: service/http.proxy/src/main/java/org/kaazing/gateway/service/http/proxy/HttpProxyServiceHandler.java [hunk 2]
```diff
@@ -179,23 +178,25 @@ private static boolean processHopByHopHeaders(HttpSession src, HttpSession dest)
...
-        for(Map.Entry<String, List<String>> e : src.getReadHeaders().entrySet()) {
+        for (Map.Entry<String, List<String>> e : src.getReadHeaders().entrySet()) {
             String name = e.getKey();
-            for(String value : e.getValue()) {
+            for (String value : e.getValue()) {
                 if (!hopByHopHeaders.contains(name)) {
-                    dest.addWriteHeader(name, value);
+                    if (name.equalsIgnoreCase(HttpHeaders.HEADER_HOST)) {
+                        dest.addWriteHeader(name, dest.getRemoteAddress().getResource().getAuthority());
+                    } else {
```
Comment (jitsni) [inline service/http.proxy/src/main/java/org/kaazing/gateway/service/http/proxy/HttpProxyServiceHandler.java]: Can you also look at HttpConnectProcessor where it adds Host header ? Also see the history w.r.t Host header and may be a fix in that file would be better. Also, I need to think a bit here. 

We may provide a config option to override the Host header. In that case, we need to see what the fix would be.

File: transport/http/src/main/java/org/kaazing/gateway/transport/http/bridge/filter/HttpResponseDecodingState.java (hunk 1 above)
Comment (msalavastru) [inline transport/http/src/main/java/org/kaazing/gateway/transport/http/bridge/filter/HttpResponseDecodingState.java]: @cmebarrow I have added the unit test that covers this fix in HttpResponseDecoderTest. I think we no longer need the k3po test, as the unit test seems sufficient, and the k3po would just be redundant.

File: service/http.proxy/src/main/java/org/kaazing/gateway/service/http/proxy/HttpProxyServiceHandler.java (hunk 2 above)
Comment (msalavastru) [inline service/http.proxy/src/main/java/org/kaazing/gateway/service/http/proxy/HttpProxyServiceHandler.java]: @jitsni I have moved the fix to HttpConnectProcessor, it seemed to be a better place for it. And also modifies only the HttpRequestMessage with the updated header and not the headers from the session. Let me know what you think.

File: transport/http/src/main/java/org/kaazing/gateway/transport/http/HttpConnectProcessor.java [hunk 3]
```diff
@@ -87,12 +87,13 @@ protected void finishConnect(DefaultHttpSession session) {
                 // TODO: strip port if default
                 httpRequest.setHeader(HttpHeaders.HEADER_HOST, resource.getAuthority());
             }
```
Comment (jitsni) [inline transport/http/src/main/java/org/kaazing/gateway/transport/http/HttpConnectProcessor.java]: I think this if statement is NOT required (as HEADER_HOST is set below unconditionally)

**Comment by jitsni** [comment]
Also, resolve the conflicts.

**Comment by msalavastru** [comment]
@jitsni I have applied the review change and resolved the conflicts.

**Pull request merged**